src/app/page.tsx

```diff
@@ -23,46 +24,73 @@ export default function Home() {
   };
 
   const handleTodoChange = async (id: number, state: ITodoItem) => {
+    const prevTodos = [...todos];
     if (state.title.trim() === "") {
       return;
     }
-    const newData = await updateData(id, state);
 
-    setTodos((todos) => todos.map((todo) => (todo.id === id ? newData : todo)));
+    setTodos((prevTodos) =>
+      prevTodos.map((todo) => (todo.id === id ? { ...todo, ...state } : todo))
+    );
+
+    try {
+      await updateData(id, state);
+    } catch (error) {
+      setTodos(prevTodos);
+    }
   };
 
   const handleAddTodo = async (todo: IAddTodoItemData) => {
     if (todo.title.trim() === "") {
```

Comment:
This is not necessary here because you're already handling this logic in the handleSubmit function